File: java/gameserverSDK/src/main/java/com/microsoft/azure/gaming/GameserverSDK.java
package com.microsoft.azure.gaming;

import java.time.ZonedDateTime;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.function.*;

/**
 * The SDK game hosts can use to integrate with PlayFab Multiplayer Servers.
 */
public class GameserverSDK {
    /**
     * Key to use in the map returned by getConfigSettings to retrieve the agent url
     */
    public static final String HEARTBEAT_ENDPOINT_KEY = "heartbeatEndpoint";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the game server instance id
     */
    public static final String SERVER_ID_KEY = "serverId";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the log folder that is uploaded to Azure blob
     * storage once the container exits
     */
    public static final String LOG_FOLDER_KEY = "logFolder";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the folder which contains all game
     * certificate files
     */
    public static final String CERTIFICATE_FOLDER_KEY = "certificateFolder";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the PlayFab title Id for this game server
     */
    public static final String TITLE_ID_KEY = "titleId";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the PlayFab build Id for this game server
     */
    public static final String BUILD_ID_KEY = "buildId";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the region this game server is running in
     */
    public static final String REGION_KEY = "region";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the session cookie passed in to this
     * game server from the allocate call. Note that this key will only appear in the map after allocation,
     * once readyForPlayers returns true.
     */
    public static final String SESSION_COOKIE_KEY = "sessionCookie";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the session id for this game server, as
     * specified in the allocate call. Note that this key will only appear in the map after allocation, once
     * readyForPlayers returns true.
     */
    public static final String SESSION_ID_KEY = "sessionId";

    /**
     * Key to use in the map returned by getConfigSettings to retrieve the vm id of the virtual machine that
     * this game server is running on.
     */
    public static final String VM_ID_KEY = "vmId";

    private static boolean isInitialized = false;

    private static HeartbeatThread heartbeatThread;

    private static final Configuration gsdkConfiguration;

    private GameserverSDK(){
        // Users should never be able to instantiate this class.
    }

    // Calling start here so that any method users call will guarantee to have the GSDK initialized
    static {
        gsdkConfiguration = new JsonFileConfiguration();
        Logger.SetLogFolder(gsdkConfiguration.getLogFolder());
        start();
    }

    /**
     * Blocks indefinitely until the Agent tells us to activate or terminate
     * @throws GameserverSDKInitializationException if our thread is interrupted while waiting
     * @return True if the server has been allocated (is about to receive players). False if the server is terminated.
     */
    public static boolean readyForPlayers() throws GameserverSDKInitializationException {
        if (heartbeatThread.getState() != SessionHostStatus.Active)
        {
            heartbeatThread.setState(SessionHostStatus.StandingBy);

            // Wait for active semaphore
            try {
                heartbeatThread.waitForTransitionToActive(0, TimeUnit.MILLISECONDS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt(); // Let caller know this thread is being interrupted
                throw new GameserverSDKInitializationException("Our thread was interrupted while waiting for Active.", e);
            }
        }

        return heartbeatThread.getState() == SessionHostStatus.Active;
    }

    /**
     * Retrieves a Map of config settings from the Agent
     * @return the configuration settings, note that these are
     * only guaranteed to be populated after the readyForPlayers()
     * call returns.
     */
    public static Map<String, String> getConfigSettings() {
        return heartbeatThread.getConfigSettings();
    }

    /**
     * Starts the heartbeat to the Agent.
     *
     * @throws GameserverSDKInitializationException if there was a problem initializing the GSDK
     */
    public static synchronized void start() {
        if (!isInitialized)
        {
            heartbeatThread = new HeartbeatThread(SessionHostStatus.Initializing, gsdkConfiguration);
            Thread heartbeat = new Thread(heartbeatThread);
            heartbeat.setDaemon(true);
            heartbeat.start();
            isInitialized = true;
        }
    }

    /**
     * Tells the agent which players are connected to this game host
     * @param players the updated list of players that are currently connected
     */
    public static void updateConnectedPlayers(List<ConnectedPlayer> players){
        heartbeatThread.setConnectedPlayers(players);
    }

    /**
     * Registers a function that will be called when the Agent tells us to Terminate
     * @param callback the function we will call on Terminate
     */
    public static void registerShutdownCallback(Runnable callback){
        heartbeatThread.registerShutdownCallback(callback);
    }

    /**
     * Registers a function that will be called when the Agent asks for our game health
     * @param callback the function we will call when we need to get the game host health
     */
    public static void registerHealthCallback(Supplier<GameHostHealth> callback){
        heartbeatThread.registerHealthCallback(callback);
    }

    /**
     * Registers a function that will be called when the Agent tells us a scheduled
     * maintenance is coming up, we will pass in the UTC Datetime of the maintenance event
     * @param callback the function we will call on a new scheduled maintenance
     */
    public static void registerMaintenanceCallback(Consumer<ZonedDateTime> callback){
        heartbeatThread.registerMaintenanceCallback(callback);
    }

    /**
     * Provides a way for the game host to log a custom message.
     * @param message the message to log
     */
    public static void log(String message) {
        Logger.Instance().Log(message);
    }

    /**
     * Returns a path to the directory where the game server can save any custom log files.
     * All files in this path will be zipped and uploaded to Azure Blob Storage once the container exits
     * @return path to the folder where all logs should be saved
     */
    public static String getLogsDirectory(){
        return gsdkConfiguration.getLogFolder();
    }

    /**
     * Returns a path to the directory shared by all game servers on the VM to store user generated content and other data.
     * @return path to the folder shared by all game servers on the VM.
     */
    public static String getSharedContentDirectory(){
        return gsdkConfiguration.getSharedContentFolder();
    }

    /**
     * Returns a path to the directory where game certificate files can be found
     * @return path to the certificate files
     */
    public static String getCertificateDirectory(){
        return gsdkConfiguration.getCertificateFolder();
    }

    /**
     * Returns the identity of the virtual machine that the games server is running on.
     * @return identity of the virtual machine.
     */
    public static String getVmId(){
        return gsdkConfiguration.getVmId();
    }

    /**
     * After allocation, returns a list of the initial players that have access to this game server,
     * used by PlayFab's Matchmaking offering
     * @return A list of player ids of the initial players that will connect
     */
    public static List<String> getInitialPlayers() { return heartbeatThread.getInitialPlayers(); }
}

File: csharp/GSDK_CSharp_Standard/GameserverSDK.cs
namespace Microsoft.Playfab.Gaming.GSDK.CSharp
{
    using Microsoft.Playfab.Gaming.GSDK.CSharp.Model;
    using System;
    using System.Collections.Generic;
    
    public enum GameState
    {
        Invalid,
        Initializing,
        StandingBy,
        Active,
        Terminating,
        Terminated,
        Quarentined
    }

    public enum GameOperation
    {
        Invalid,
        Continue,
        Active,
        Terminate
    }

    public static class GameserverSDK
    {
        // These two keys are only available after allocation (once readyForPlayers returns true)
        public const string SessionCookieKey = "sessionCookie";
        public const string SessionIdKey = "sessionId";
        public const string HeartbeatEndpointKey = "heartbeatEndpoint";
        public const string ServerIdKey = "serverId";
        public const string LogFolderKey = "logFolder";
        public const string SharedContentFolderKey = "sharedContentFolder";
        public const string CertificateFolderKey = "certificateFolder";
        public const string TitleIdKey = "titleId";
        public const string BuildIdKey = "buildId";
        public const string RegionKey = "region";
        public const string VmIdKey = "vmId";
        public const string PublicIpV4AddressKey = "publicIpV4Address";
        public const string FullyQualifiedDomainNameKey = "fullyQualifiedDomainName";

        public const string GsdkConfigFileEnvVarKey = "GSDK_CONFIG_FILE";

        private static readonly InternalSdk _internalSdk = new InternalSdk();

        /// <summary>
        /// Called when the game server is ready to accept clients.
        /// If Start() hasn't been called by this point, it will be
        /// called implicitly here.
        /// </summary>
        /// <remarks>Required</remarks>
        /// <returns>
        /// True if the game server was allocated (clients are about to connect).
        /// False if the game server was terminated and is about to be shut down.
        /// </returns>
        public static bool ReadyForPlayers()
        {
            _internalSdk.Start();

            if (_internalSdk.State != GameState.Active)
            {
                _internalSdk.TransitionToActiveEvent.Reset();
                _internalSdk.State = GameState.StandingBy;
                _internalSdk.TransitionToActiveEvent.WaitOne();
            }

            return _internalSdk.State == GameState.Active;
        }

        /// <summary>
        /// Gets information (ipAddress and ports) for connecting to the game server, as well as the ports the
        /// game server should listen on.
        /// </summary>
        /// <returns></returns>
        public static GameServerConnectionInfo GetGameServerConnectionInfo()
        {
            _internalSdk.Start();
            return _internalSdk.GetGameServerConnectionInfo();
        }

        /// <summary>
        /// Returns all configuration settings
        /// </summary>
        /// <returns>Optional</returns>
        public static IDictionary<string, string> getConfigSettings()
        {
            _internalSdk.Start();

            return new Dictionary<string, string>(_internalSdk.ConfigMap, StringComparer.OrdinalIgnoreCase);
        }

        /// <summary>
        /// Kicks off communication threads, heartbeats, etc.
        /// Called implicitly by ReadyForPlayers if not called beforehand.
        /// </summary>
        public static void Start(bool debugLogs = false)
        {
            _internalSdk.Start(debugLogs);
        }

        /// <summary>
        /// Tells the PlayFab service information on who is connected.
        /// </summary>
        /// <param name="currentlyConnectedPlayers"></param>
        public static void UpdateConnectedPlayers(IList<ConnectedPlayer> currentlyConnectedPlayers)
        {
            _internalSdk.Start();
            _internalSdk.ConnectedPlayers = currentlyConnectedPlayers;
        }

        /// <summary>
        /// Gets called if the server is shutting us down
        /// </summary>
        /// <param name="callback">The callback</param>
        public static void RegisterShutdownCallback(Action callback)
        {
            _internalSdk.Start();

            _internalSdk.ShutdownCallback = callback;
        }

        /// <summary>
        /// Gets called when the agent needs to check on the
        /// game's health
        /// </summary>
        /// <param name="callback">The callback</param>
        public static void RegisterHealthCallback(Func<bool> callback)
        {
            _internalSdk.Start();

            _internalSdk.HealthCallback = callback;
        }

        /// <summary>
        /// DEPRECATED Use RegisterMaintenanceV2Callback instead.
        /// Gets called if the server is getting a scheduled maintenance,
        /// it will get the UTC Datetime of the maintenance event as an argument.
        /// </summary>
        /// <param name="callback">The callback</param>
        [Obsolete]
        public static void RegisterMaintenanceCallback(Action<DateTimeOffset> callback)
        {
            _internalSdk.Start();

            _internalSdk.MaintenanceCallback = callback;
        }

        /// <summary>
        /// Gets called if the server is getting a scheduled maintenance,
        /// it will get the maintenance event as an argument.
        /// </summary>
        /// <remarks>
        /// https://learn.microsoft.com/azure/virtual-machines/windows/scheduled-events#event-properties
        /// </remarks>
        /// <param name="callback">The callback</param>
        public static void RegisterMaintenanceV2Callback(Action<MaintenanceSchedule> callback)
        {
            _internalSdk.Start();

            _internalSdk.MaintenanceV2Callback = callback;
        }

        /// <summary>
        /// Returns the directory whose contents will be uploaded so logs
        /// can be easily retrieved
        /// </summary>
        /// <returns>A path to the directory to place logs in</returns>
        public static string GetLogsDirectory()
        {
            _internalSdk.Start();

            if (_internalSdk.ConfigMap.TryGetValue(LogFolderKey, out string folder))
            {
                return folder;
            }

            return string.Empty;
        }

        /// <summary>
        /// Returns the directory whose contents are shared among all game servers within a VM.
        /// </summary>
        /// <returns>A path to the directory where shared content can be stored (temporarily). </returns>
        public static string GetSharedContentDirectory()
        {
            _internalSdk.Start();

            return _internalSdk.ConfigMap.TryGetValue(SharedContentFolderKey, out string folder) ? folder : string.Empty;
        }

        /// <summary>
        /// After allocation, returns a list of the initial players that have access to this game server, used by PlayFab's
        /// Matchmaking offering
        /// </summary>
        /// <returns>A list of player ids of the initial players that will connect</returns>
        public static IList<string> GetInitialPlayers()
        {
            _internalSdk.Start();
        
            return new List<string>(_internalSdk.InitialPlayers);
        }

        /// <summary>
        /// Adds a custom log message to the GSDK log output
        /// </summary>
        /// <param name="message">The message to be logged</param>
        public static void LogMessage(string message)
        {
            _internalSdk.Start();
            _internalSdk.Logger.Log(message);
        }
    }
}